gax/src/main/java/com/google/api/gax/batching/v2/FlowController.java

```diff
@@ -0,0 +1,215 @@
+    }
+
+    public long getCurrentMaxBatchBytes() {
+      return currentMaxBytes;
+    }
+
+    @Override
+    public String toString() {
+      return String.format(
+          "The maximum number of batch bytes: %d have been reached.", currentMaxBytes);
+    }
+  }
+
+  /**
+   * Enumeration of behaviors that FlowController can use in case the flow control limits are
+   * exceeded.
+   */
+  @BetaApi
+  public enum LimitExceededBehavior {
+    /**
+     * Throws {@link MaxOutstandingElementCountReachedException} or {@link
+     * MaxOutstandingRequestBytesReachedException}.
+     *
+     * <p>This might be appropriate in interactive scenarios. For example, a web server might catch
+     * these exceptions and report to the user that the system is overloaded and that the user could
+     * try again later. It could also be useful in applications that implement custom rate-limiting
+     * logic.
+     */
+    ThrowException,
+
+    /**
+     * Waits until the request can be made without exceeding the limit.
+     *
+     * <p>This might be appropriate in batch-processing, where latencies of individual requests are
+     * not important.
+     */
+    Block,
+
+    /**
+     * Disables flow-control.
+     *
+     * <p>This is provided mainly for debugging and not recommended for production use. Having too
+     * many requests in-flight might cause RPCs to fail due to congested network or the computer to
+     * run out of memory due to excessive buffering, etc.
+     */
+    Ignore,
+  }
+
+  @Nullable private final Semaphore outstandingElementCount;
+  @Nullable private final Semaphore outstandingByteCount;
+  private final int maxOutstandingElementCount;
+  private final long maxOutstandingRequestBytes;
+  private final boolean isBlocking;
+
+  public FlowController(FlowControlSettings settings) {
+    isBlocking = settings.getLimitExceededBehavior() == LimitExceededBehavior.Block;
+    switch (settings.getLimitExceededBehavior()) {
+      case ThrowException:
+      case Block:
+        break;
+      case Ignore:
+        this.maxOutstandingElementCount = 0;
+        this.maxOutstandingRequestBytes = 0;
+        this.outstandingElementCount = null;
+        this.outstandingByteCount = null;
+        return;
+      default:
+        throw new IllegalArgumentException(
+            "Unknown LimitBehaviour: " + settings.getLimitExceededBehavior());
+    }
+
+    this.maxOutstandingElementCount = settings.getMaxOutstandingElementCount();
+    if (maxOutstandingElementCount == 0) {
+      outstandingElementCount = null;
+    } else {
+      outstandingElementCount = new Semaphore(maxOutstandingElementCount, true);
+    }
+
+    this.maxOutstandingRequestBytes = settings.getMaxOutstandingRequestBytes();
+    if (maxOutstandingRequestBytes == 0) {
+      outstandingByteCount = null;
+    } else {
+      int maxRequestBytes = Ints.checkedCast(maxOutstandingRequestBytes);
+      outstandingByteCount = new Semaphore(maxRequestBytes, true);
```

Comment:
Please remove the cast & use Semaphore64

Reply:
Thanks for the suggestion, I have updated rquestByteCount with `Semaphore64`. Please have a fresh look.